cns/deviceplugin/plugin.go

```diff
@@ -0,0 +1,143 @@
+func (p *Plugin) Run(ctx context.Context) {
+	defer p.mustCleanUp()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			p.Logger.Info("starting device plugin for resource", zap.String("resource", p.ResourceName))
+			if err := p.run(ctx); err != nil {
+				p.Logger.Error("device plugin for resource exited", zap.String("resource", p.ResourceName), zap.Error(err))
+			}
+		}
+	}
+}
+
+func (p *Plugin) run(ctx context.Context) error {
+	childCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	s := NewServer(p.Socket, p.Logger, p, p.deviceCheckInterval)
+	runErrChan := make(chan error, 2) //nolint:gomnd // disabled in favor of readability
+	go func(errChan chan error) {
+		if err := s.Run(childCtx); err != nil {
+			errChan <- err
+		}
+	}(runErrChan)
+
+	// wait till the server is ready before registering with kubelet
+	readyErrChan := make(chan error, 2) //nolint:gomnd // disabled in favor of readability
+	go func(errChan chan error) {
+		errChan <- s.Ready(childCtx)
+	}(readyErrChan)
+
+	select {
+	case err := <-runErrChan:
+		return errors.Wrap(err, "error starting grpc server")
+	case err := <-readyErrChan:
+		if err != nil {
+			return errors.Wrap(err, "error waiting on grpc server to be ready")
+		}
+	case <-ctx.Done():
+		return nil
+	}
+
+	p.Logger.Info("registering with kubelet")
+	// register with kubelet
+	if err := p.registerWithKubelet(childCtx); err != nil {
+		return errors.Wrap(err, "failed to register with kubelet")
+	}
+
+	// run until the socket goes away or the context is cancelled
+	<-p.SocketWatcher.WatchSocket(childCtx, p.Socket)
+	return nil
+}
+
+func (p *Plugin) registerWithKubelet(ctx context.Context) error {
+	conn, err := grpc.Dial(p.kubeletSocket, grpc.WithTransportCredentials(insecure.NewCredentials()), //nolint:staticcheck // TODO: Move to grpc.NewClient method
+		grpc.WithContextDialer(func(ctx context.Context, addr string) (net.Conn, error) {
+			d := &net.Dialer{}
+			conn, err := d.DialContext(ctx, "unix", addr)
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to dial context")
+			}
+			return conn, nil
+		}))
+	if err != nil {
+		return errors.Wrap(err, "error connecting to kubelet")
+	}
+	defer conn.Close()
+
+	client := v1beta1.NewRegistrationClient(conn)
+	request := &v1beta1.RegisterRequest{
+		Version:      v1beta1.Version,
+		Endpoint:     filepath.Base(p.Socket),
+		ResourceName: p.ResourceName,
+	}
+	if _, err = client.Register(ctx, request); err != nil {
+		return errors.Wrap(err, "error sending request to register with kubelet")
+	}
+	return nil
+}
+
+func (p *Plugin) mustCleanUp() {
+	p.Logger.Info("cleaning up device plugin")
+	if err := os.Remove(p.Socket); err != nil && !os.IsNotExist(err) {
+		p.Logger.Panic("failed to remove socket", zap.Error(err))
+	}
+}
+
+func (p *Plugin) UpdateDeviceCount(count int) {
+	p.deviceCountMutex.Lock()
+	defer p.deviceCountMutex.Unlock()
+	p.deviceCount = count
```

Comment:
for a small write-only block like this, prefer to directly unlock. it's lighter and just as readable
```suggestion
	p.deviceCount = count
	p.deviceCountMutex.Unlock()
```

Reply:
Actually, do these need to be synchronized? it's primitive `int`, it has separate setter/getter methods, it is not directly accessed other than here...
I think the lock here is not necessary? WDYT @timraymond?

Reply:
if removed here we probably want to move it to `TrackDevices` which is probably where it should have been all along

Reply:
I think the lock is necessary. It's not the way it's accessed that's the issue... it's that the write could occur to a cache on one core that would not be visible to another thread on another core. It _could_ be okay in some execution histories, but not in arbitrary sequencing of events.

Also, for the defer... idk. For this trivial usage, yeah, you can drop the defer. I still kinda like it though, since it will run even in the presence of panics. Yes, this can't panic, but leaving the defer hardens it against future changes where the author was not so careful auditing for panics. I'd rather have a fatter critical section than "forget" to drop a lock because we panicked and it was recovered somewhere up the call stack (and we kept going). I feel similarly about Wait Groups and `Done()`. Really, what I'd like is something like:

```go
err := mu.WithLock(func() error {
       // some critical section
})
```

which some languages do... and we could Greenspun that feature, but it would be weird to introduce.

Reply:
> for a small write-only block like this, prefer to directly unlock. it's lighter and just as readable

Updated